### ISSUE TITLE
buf protoc command line option inconsistency with protoc

### ISSUE THREAD

**Issue by bufdev** [bufbuild/buf#491]
```shell
$ buf protoc \
    --go_out ../gen/proto \
    --go_opt plugins=grpc \
    --go_opt Macme/payment/v2/payment.proto=demo/consumer/gen/proto/acme/payment/v2 \
    $(find . -type f -name "*.proto")
```
This results in the error:
```
duplicate --go_out
```

However,
```shell
$ protoc \
    --go_out ../gen/proto \
    --go_opt plugins=grpc \
    --go_opt Macme/payment/v2/payment.proto=demo/consumer/gen/proto/acme/payment/v2 \
    $(find . -type f -name "*.proto")
```
works fine.

Interestingly,
```shell
$ protoc $(find . -type f -name "*.proto") \
    --go_out ../gen/proto \
    --go_opt plugins=grpc \
    --go_opt Macme/payment/v2/payment.proto=demo/consumer/gen/proto/acme/payment/v2
```
Produces the same `duplicate --go_out` error in `protoc` as in `buf protoc`.

**Comment by Houndie** [comment]
I did a little bit of research into this, enough to figure out the gist of the problem, not enough to figure out how you want to solve it.

So a few notes:  I couldn't replicate the "interestingly" note.  The final command above works for me. :shrug: 

Secondly, I was able to reduce your reproducer down to a much simpler form:

```
buf protoc --go_out . routeguide/route_guide.proto
```

Which is obviously distressing that that doesn't even work.  However, if I add an equals sign:

```
buf protoc --go_out=. routeguide/route_guide.proto
```

This works fine, which lead me to the problem.

We are counting the number of times a specific `_out` flag is called [here](https://github.com/bufbuild/buf/blob/0fb4800758f7fd6bf9637e5660c6f63b409eed0a/private/buf/cmd/buf/command/protoc/flags.go#L258) and using it [here](https://github.com/bufbuild/buf/blob/main/private/buf/cmd/buf/command/protoc/flags.go#L388) to return an error if this flag was provided too many times.

The problem is that this value is being generated by the `pflags` `Normalize` function, and, well, that's just not an accurate way to count the number of times a flag appears.  A quick dip into the [`pflags` library](https://github.com/spf13/pflag/blob/v1.0.5/flag.go) shows that flags are normalized all the time, and if you look into cobra, you'll find that cobra touches these functions quite a bit as well.  In this case, you'll find that cobra is actually causing the normalize function to be called an additional 4 times when you're given an argument with a space instead of an equals sign.

As for a solution, that's a bit more speculative.  I should note that protoc (3.17.3) doesn't actually care about duplicate go_out flags as long as the values are different, so you may just be able to delete this count and move on.  If you want to move to a more pflagy solution, that's a bit tough...pflags doesn't handle any kind of regular expression flag matching, which I'm sure is why this hack was necessary to begin with.  I'm thinking it might be possible to bind a new flag in `Normalize` so that you can bind "go_out" and "js_out" and "whatever_out" as they show up dynamically (as long as care is not taken to recurse infinitely, as `Normalize` is going to be called from `BindFlag`).  The other option would be to iterate through the `args` array, search for anything that begins with `--` and ends with `_out` and bind that flag before flags are parsed.

**Comment by Houndie** [comment]
Playing with `protoc` to figure out the current behavior, here's what we got:

Given multiple identical `_out` flag, `protoc` will respect them as if they were different out parameters:

```
protoc --go_out=a --go_out=b route_guide/routeguide/route_guide.proto #generates files in both a/ and b/
```

Options provided as part of the `_out` flag are specific to that `_out` paramemter

```
protoc --go_out=paths=source_relative:a --go_out=b route_guide/routeguide/route_guide.proto #only files in a/ are generated with source_relative
```

Options provided as part of the `_opt` flag are applied to every invocation of the matching plugin

```
protoc --go_out=a --go_out=b --go_opt=source_relative route_guide/routeguide/route_guide.proto #both a/ and b/ are generated with source_relative
```

It looks like the `env` struct inside `buf protoc` is designed around only having one invocation of each type of a plugin so a small refactor might be necessary if feature parity between `protoc` and `buf protoc` is desired.

**Comment by bufdev** [comment]
We have now deleted `buf protoc` by moving it to alpha in #915, and will likely eventually delete it entirely. As such, going to close this issue as `buf protoc` will almost for sure not be worked on in the future. If we decide to revisit this decision, we should reopen this issue. All issues that were closed are tracked on the #915.
